lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/ApiSpec.hs

```diff
@@ -10,23 +12,135 @@ module Cardano.Wallet.Jormungandr.ApiSpec
+    ( forM_, replicateM )
 import Data.Aeson
     ( eitherDecode )
+import Data.Aeson.QQ
+    ( aesonQQ )
+import Data.Proxy
+    ( Proxy (..) )
 import Data.Quantity
     ( Quantity (..) )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( FromText (..), TextDecodingError (..), ToText (..) )
+import Data.Word
+    ( Word64 )
+import Test.Aeson.Internal.RoundtripSpecs
+    ( roundtripSpecs )
 import Test.Hspec
     ( Spec, describe, it, shouldBe )
+import Test.QuickCheck
+    ( Arbitrary (..) )
+import Test.Text.Roundtrip
+    ( textRoundtrip )
 
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString as BS
 import qualified Data.ByteString.Lazy as BL
 
 spec :: Spec
 spec = do
     describe "Jormungandr Api" $ do
 
+        describe "Textual roundtrip tests for API types" $ do
+
+            textRoundtrip $ Proxy @AccountId
+
+        describe "JSON roundtrip tests for API types" $ do
+
+            roundtripSpecs $ Proxy @AccountState
+
+        it "Valid account IDs are properly decoded from text" $ do
+
+            forM_ testAccountIdTexts $ \text ->
+                toText <$> fromText @AccountId text
+                    `shouldBe` Right text
+
+        it "Invalid account IDs cannot be decoded from text" $ do
+
+            let invalidAccountIdTexts =
+                    [ ""
+                    , "a"
+                    , "0123456789abcdef"
+                    ]
+            let expectedErrorMessage =
+                    "Invalid Jörmungandr account ID: \
+                    \expecting a hex-encoded value that is 32 bytes in length."
+            forM_ invalidAccountIdTexts $ \text ->
+                toText <$> fromText @AccountId text
+                    `shouldBe` Left (TextDecodingError expectedErrorMessage)
+
+        describe "Example account state objects are properly decoded" $ do
+
+            let testBalance = 1000
+            let testTransactionCount = 1
+            let testPoolRatio1 = 1
+            let testPoolRatio2 = 10
+
+            it "With 0 stake pools" $ do
+
+                let testAccountState = Aeson.encode [aesonQQ|
+                        { "value": #{testBalance}
+                        , "counter": #{testTransactionCount}
+                        , "delegation": {"pools": []}
+                        }|]
+                eitherDecode testAccountState `shouldBe` Right AccountState
+                    { currentBalance = Quantity testBalance
+                    , totalTransactionCount = Quantity testTransactionCount
+                    , stakePools = []
+                    }
+
+            it "With 1 stake pool" $ do
+
+                let testAccountState = Aeson.encode [aesonQQ|
+                        { "value": #{testBalance}
+                        , "counter": #{testTransactionCount}
+                        , "delegation":
+                            { "pools":
+                                [[#{toText testPoolId1}, #{testPoolRatio1}]]
```

Comment:
I wonder if this test couldn't be better if we didn't use `toText`. 

It wouldn't catch the instances changing.

If we hard-coded the hex string here we would guarantee that we're encoding/decoding hex-correctly. We could also use `unsafeFromHex`/`unsafeToHex`.

I also wonder how this test relates to the aeson roundtrip.